### ISSUE TITLE
[BUG] Script ran too long

### ISSUE THREAD

**Issue by zaphon** [teelolws/TLDRMissions#58]
Crashed calculating..  Mission is Spines of Steel and Sorrow

`1x TLDRMissions\simulation.lua:771: script ran too long
[string "@TLDRMissions\simulation.lua"]:771: in function <TLDRMissions\simulation.lua:719>
[string "@TLDRMissions\simulation.lua"]:901: in function <TLDRMissions\simulation.lua:875>
[string "@TLDRMissions\simulation.lua"]:1127: in function <TLDRMissions\simulation.lua:1053>
[string "@TLDRMissions\simulation.lua"]:1479: in function <TLDRMissions\simulation.lua:1423>
[string "@TLDRMissions\simulation.lua"]:1466: in function <TLDRMissions\simulation.lua:1423>
[string "@TLDRMissions\simulation.lua"]:1488: in function <TLDRMissions\simulation.lua:1423>
[string "@TLDRMissions\simulation.lua"]:1488: in function <TLDRMissions\simulation.lua:1423>
[string "@TLDRMissions\simulation.lua"]:1466: in function <TLDRMissions\simulation.lua:1423>
[string "@TLDRMissions\simulation.lua"]:1488: in function <TLDRMissions\simulation.lua:1423>
[string "@TLDRMissions\simulation.lua"]:1488: in function <TLDRMissions\simulation.lua:1423>
[string "@TLDRMissions\simulation.lua"]:1466: in function <TLDRMissions\simulation.lua:1423>
[string "@TLDRMissions\simulation.lua"]:1488: in function <TLDRMissions\simulation.lua:1423>
...
[string "@TLDRMissions\simulation.lua"]:1488: in function <TLDRMissions\simulation.lua:1423>
[string "@TLDRMissions\simulation.lua"]:1488: in function <TLDRMissions\simulation.lua:1423>
[string "@TLDRMissions\simulation.lua"]:1466: in function <TLDRMissions\simulation.lua:1423>
[string "@TLDRMissions\simulation.lua"]:1488: in function <TLDRMissions\simulation.lua:1423>
[string "@TLDRMissions\simulation.lua"]:1488: in function <TLDRMissions\simulation.lua:1423>
[string "@TLDRMissions\simulation.lua"]:1466: in function `turnRecursion'
[string "@TLDRMissions\simulation.lua"]:1495: in function <TLDRMissions\simulation.lua:27>
[string "@TLDRMissions\simulation.lua"]:1658: in function `Simulate'
[string "@TLDRMissions\combinations.lua"]:7: in function `func'
[string "@TLDRMissions\workhandler.lua"]:38: in function <TLDRMissions\workhandler.lua:23>

Locals:
attacker = <table> {
 maxHP = 3390
 spells = <table> {
 }
 unregisterPassive = <function> defined @TLDRMissions\simulation.lua:1219
 autoAttack = <table> {
 }
 boardIndex = 7
 name = "Steelspine Vanguard"
 followerID = -2
 baseAttack = 77
 buffs = <table> {
 }
 HP = 270
 thorns = <table> {
 }
}
defender = <table> {
 maxHP = 900
 spells = <table> {
 }
 autoAttack = <table> {
 }
 boardIndex = 3
 name = "Venthyr Nightblade"
 followerID = "0xFFFFFFFFFFFFFFFF"
 baseAttack = 108
 buffs = <table> {
 }
 HP = 595
 thorns = <table> {
 }
}
attackPercent = 100
changeDamageTakenRaw = <table> {
 lowestID = 999999
 action = <function> defined @TLDRMissions\simulation.lua:758
 amount = 0
}
changeDamageDealtRaw = <table> {
 lowestID = 999999
 action = <function> defined @TLDRMissions\simulation.lua:757
 amount = 0
}
changeDamageTakenPercent = <table> {
 lowestID = 999999
 action = <function> defined @TLDRMissions\simulation.lua:760
 amount = 100
}
changeDamageDealtPercent = <table> {
 lowestID = 999999
 action = <function> defined @TLDRMissions\simulation.lua:759
 amount = 100
}
damage = 77
actions = <table> {
 1 = <table> {
 }
 2 = <table> {
 }
 3 = <table> {
 }
 4 = <table> {
 }
}
sort_func = <function> defined @TLDRMissions\simulation.lua:764
(for generator) = <function> defined =[C]:-1
(for state) = <table> {
 1 = <table> {
 }
 2 = <table> {
 }
 3 = <table> {
 }
 4 = <table> {
 }
}
(for control) = 1
_ = 1
action = <table> {
 lowestID = 999999
 action = <function> defined @TLDRMissions\simulation.lua:758
 amount = 0
}
(*temporary) = 0
(*temporary) = 1
(*temporary) = <table> {
 lowestID = 999999
 action = <function> defined @TLDRMissions\simulation.lua:758
 amount = 0
}
(*temporary) = false
(*temporary) = "script ran too long"
field = <table> {
 1 = <table> {
 }
 2 = <table> {
 }
 3 = <table> {
 }
 4 = <table> {
 }
 5 = <table> {
 }
 6 = <table> {
 }
 7 = <table> {
 }
 8 = <table> {
 }
 9 = <table> {
 }
 10 = <table> {
 }
 11 = <table> {
 }
}
`

**Comment by teelolws** [comment]
I uploaded a patch about ~10 minutes ago. Can you update and try it again?

**Comment by zaphon** [comment]
Already skipped the mission and moved on. But I grabbed it and will let you know if I run into any issues on another character (this was #4 of 25)

**Comment by zaphon** [comment]
Another issue - Mission Defensive Trials  It had crashed once after I had skipped a mission, so I logged out and back in to start fresh and got same error.

`2x TLDRMissions\guihandlers.lua:209: attempt to concatenate local 'itemLink' (a nil value)
[string "@TLDRMissions\guihandlers.lua"]:209: in function <TLDRMissions\guihandlers.lua:171>
[string "@TLDRMissions\combinations.lua"]:356: in function `callback'
[string "@TLDRMissions\combinations.lua"]:19: in function `func'
[string "@TLDRMissions\workhandler.lua"]:38: in function <TLDRMissions\workhandler.lua:23>

Locals:
results = <table> {
 defeats = 0
 missionID = 2330
 combination = <table> {
 }
 incompletes = 0
 victories = 1
}
dontClear = nil
rewards = <table> {
 1 = <table> {
 }
 2 = <table> {
 }
}
text = ""
(for generator) = <function> defined =[C]:-1
(for state) = <table> {
 1 = <table> {
 }
 2 = <table> {
 }
}
(for control) = 1
_ = 1
reward = <table> {
 quantity = 1
 itemID = 187569
 itemLink = "|cffffffff|Hitem:187569::::::::60:73::31:::::::|h[]|h|r"
}
_ = nil
itemLink = nil
(*temporary) = ""
(*temporary) = nil
(*temporary) = " x1; "
(*temporary) = "1"
(*temporary) = "; "
(*temporary) = "attempt to concatenate local 'itemLink' (a nil value)"
missionWaitingUserAcceptance = <table> {
 defeats = 0
 missionID = 2330
 combination = <table> {
 }
 incompletes = 0
 victories = 1
}
calculatedMissionBacklog = <table> {
}
alreadyUsedFollowers = <table> {
 -9223372036854775808 = true
 612032744 = true
 543861178 = true
 543860862 = true
}
addon = <table> {
 createWorkBatch = <function> defined @TLDRMissions\workhandler.lua:80
 arrangeAllFollowerPositions = <function> defined @TLDRMissions\combinations.lua:28
 reputationCurrencyIDs = <table> {
 }
 updateRewards = <function> defined @TLDRMissions\guihandlers.lua:19
 Simulate = <function> defined @TLDRMissions\simulation.lua:1503
 logSentMission = <function> defined @TLDRMissions\logging.lua:5
 isCurrentWorkBatchEmpty = <function> defined @TLDRMissions\workhandler.lua:98
 unpauseWorker = <function> defined @TLDRMissions\workhandler.lua:72
 GetAllMissions = <function> defined @TLDRMissions\missions.lua:179
 GetAugmentRuneMissions = <function> defined @TLDRMissions\missions.lua:167
 followerXPItemIDs = <table> {
 }
 GetCraftingCacheMissions = <function> defined @TLDRMissions\missions.lua:175
 GetFollowerXPItemMissions = <function> defined @TLDRMissions\missions.lua:155
 arrangeFollowerCombinationsByFewestFollowersPlusTroops = <function> defined @TLDRMissions\combinations.lua:343
 TestExistingMission = <function> defined @TLDRMissions\testing.lua:69
 craftingMaterialsCacheIDs = <table> {
 }
 GetAllMissionsMatchingFilter = <function> defined @TLDRMissions\missions.lua:67
 getCurrentWorkBatchSize = <function> defined @TLDRMissions\workhandler.lua:102
 TestNextCompletedMission = <function> defined @TLDRMissions\testing.lua:4
 clearWork = <function> defined @TLDRMissions\workhandler.lua:91
 GetGoldMissions = <function> defined @TLDRMissions\missions.lua:147
 addWork = <function> defined @TLDRMissions\workhandler.lua:86
 GetAnimaMissions = <function> defined @TLDRMissions\missions.lua:159
 SimulateFromLog = <function> defined @TLDRMissions\simulation.lua:1665
 spellsDB = <table> {
 }
 pauseWorker = <function> defined @TLDRMissions\workhandler.lua:68
 dontWait = false
 GetPetCharmMissions = <function> defined @TLDRMissions\missions.lua:163
 isWorkerPaused = <function> defined @TLDRMissions\workhandler.lua:76
 sortFollowers = <function> defined @TLDRMissions\combinations.lua:130
 TestFirstTenFollowersPlusTroops = <function> defined @TLDRMissions\testing.lua:49
 TestMission = <function> defined @TLDRMissions\testing.lua:78
 GUI = TLDRMissionsFrame {
 }
 logCompletedMission = <function> defined @TLDRMissions\logging.lua:122
 TestGetMissions = <function> defined @TLDRMissions\testing.lua:65
 GetReputationMissions = <function> defined @TLDRMissions\missions.lua:171
 registerWorkStepCallback = <function> defined @TLDRMissions\workhandler.lua:19
 arrangeFollowerCombinationsByMostFollowersPlusTroops = <function> defined @TLDRMissions\combinations.lua:157
 GetFollowerXPMissio`

**Comment by zaphon** [comment]
Getting this on every mission on this character now. It fills in the 5 actors and than throws the above error when it should be displaying the rewards, and I can't send them on the mission.

**Comment by teelolws** [comment]
Strange. Issues with itemLinks are usually a one-time thing until your game downloads the item details into the cache.

Can you try this?
[guihandlers.zip](https://github.com/teelolws/TLDRMissions/files/8536109/guihandlers.zip)

**Comment by zaphon** [comment]
It skipped it with the above file.

**Comment by teelolws** [comment]
For your first issue, I finally got a character with that mission.

![image](https://user-images.githubusercontent.com/49792789/164675628-1e919d63-db69-4c1c-94c4-0844a7f227aa.png)

I've made some changes to the "Random Ally" logic that should stop "script ran too long" errors. The downside: it'll take longer to compute. Its testing every possible target for those "Random Ally"s. Depending on how important that 317 gold is to you, you might prefer to just skip it.

I'll be doing some more testing with my changes, if I don't have any issues it'll go in tomorrows patch.
